File: 47_Permutations II/Solution.py
from typing import List

class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        def backtrack(path, used):
            if len(path) == len(nums):
                result.append(path[:])
                return
            for i in range(len(nums)):
                if used[i] or (i > 0 and nums[i] == nums[i - 1] and not used[i - 1]):
                    continue
                used[i] = True
                path.append(nums[i])
                backtrack(path, used)
                path.pop()
                used[i] = False

        nums.sort()
        result = []
        backtrack([], [False] * len(nums))
        return result

# Test the code
sol = Solution()
nums1 = [1, 1, 2]
nums2 = [1, 2, 3]
print(sol.permuteUnique(nums1))
print(sol.permuteUnique(nums2))

File: 47_Permutations II/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution { // Changed class name to Solution

    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);  // Sort the array to handle duplicates
        backtrack(result, new ArrayList<>(), nums, new boolean[nums.length]);
        return result;
    }

    private void backtrack(List<List<Integer>> result, List<Integer> tempList, int[] nums, boolean[] used) {
        if (tempList.size() == nums.length) {
            result.add(new ArrayList<>(tempList));
        } else {
            for (int i = 0; i < nums.length; i++) {
                if (used[i] || (i > 0 && nums[i] == nums[i - 1] && !used[i - 1])) continue;  // Skip duplicates
                used[i] = true;
                tempList.add(nums[i]);
                backtrack(result, tempList, nums, used);
                used[i] = false;
                tempList.remove(tempList.size() - 1);
            }
        }
    }

    public static void main(String[] args) {
        Solution sol = new Solution(); // Changed instance creation to Solution
        int[] nums1 = {1, 1, 2};
        int[] nums2 = {1, 2, 3};

        System.out.println(sol.permuteUnique(nums1));
        System.out.println(sol.permuteUnique(nums2));
    }
}
